File: server/src/services/game.service.ts
/**
 * 游戏轮次管理服务
 * 使用内存 + JSON 文件持久化（轻量方案，无需 SQLite）
 * 负责游戏状态机: 创建轮次 -> 开始 -> 接收猜测 -> 结束
 */
import fs from 'node:fs';
import path from 'node:path';
import { calculateSimilarity, isCorrect, getEmbedding, loadCachedEmbeddings } from './embedding.service.js';
import { addGuess, getLeaderboard, getGuessCount, incrementGuessCount } from './rank.service.js';
import type { LeaderboardEntry } from './rank.service.js';

// --- 类型 ---
export type RoundStatus = 'waiting' | 'active' | 'finished';

export interface Word {
  id: number;
  word: string;
  hint: string;
  category: string;
  difficulty: number;
  embedding?: number[];
}

export interface Round {
  id: number;
  wordId: number;
  word: string;
  hint: string;
  status: RoundStatus;
  startedAt: string | null;
  endedAt: string | null;
  duration: number;
}

export interface GuessRecord {
  roundId: number;
  userId: string;
  username: string;
  avatar: string;
  guessText: string;
  similarity: number;
  isCorrect: boolean;
  createdAt: string;
}

export interface GuessResult {
  similarity: number;
  isCorrect: boolean;
  rank: number;
  message: string;
}

// --- 数据存储 ---
const DATA_DIR = path.resolve(__dirname, '../../data');
const WORDS_FILE = path.join(DATA_DIR, 'words.json');
const GUESSES_FILE = path.join(DATA_DIR, 'guesses.json');

let words: Word[] = [];
let rounds: Round[] = [];
let guesses: GuessRecord[] = [];
let nextWordId = 1;
let nextRoundId = 1;

function ensureDataDir(): void {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }
}

function loadWords(): void {
  ensureDataDir();
  if (fs.existsSync(WORDS_FILE)) {
    try {
      const data = JSON.parse(fs.readFileSync(WORDS_FILE, 'utf-8'));
      words = data.words || [];
      nextWordId = data.nextId || (words.length > 0 ? Math.max(...words.map(w => w.id)) + 1 : 1);
    } catch {
      words = [];
    }
  }
}

function saveWords(): void {
  ensureDataDir();
  fs.writeFileSync(WORDS_FILE, JSON.stringify({ words, nextId: nextWordId }, null, 2));
}

function loadGuesses(): void {
  ensureDataDir();
  if (fs.existsSync(GUESSES_FILE)) {
    try {
      guesses = JSON.parse(fs.readFileSync(GUESSES_FILE, 'utf-8'));
    } catch {
      guesses = [];
    }
  }
}

function saveGuesses(): void {
  ensureDataDir();
  fs.writeFileSync(GUESSES_FILE, JSON.stringify(guesses));
}

/**
 * 初始化: 加载词库 + embedding 缓存
 */
export function initGame(): void {
  loadWords();
  loadGuesses();

  // 加载已有 embedding 缓存
  const cachedRows = words
    .filter(w => w.embedding)
    .map(w => ({ word: w.word, embedding: JSON.stringify(w.embedding) }));
  loadCachedEmbeddings(cachedRows);

  console.log(`[Game] 初始化完成，词库 ${words.length} 个词`);
}

/**
 * 创建新一轮游戏
 */
export async function createRound(options?: {
  wordId?: number;
  category?: string;
  difficulty?: number;
  duration?: number;
}): Promise<Round> {
  const duration = options?.duration || parseInt(process.env.ROUND_DURATION || '60');

  let word: Word;

  if (options?.wordId) {
    word = words.find(w => w.id === options.wordId)!;
    if (!word) throw new Error(`词 ID ${options.wordId} 不存在`);
  } else {
    // 按条件筛选后随机
    let candidates = words;
    if (options?.category) candidates = candidates.filter(w => w.category === options.category);
    if (options?.difficulty) candidates = candidates.filter(w => w.difficulty === options.difficulty);
    if (candidates.length === 0) throw new Error('词库中没有符合条件的词');
    word = candidates[Math.floor(Math.random() * candidates.length)];
  }

  // 预计算 embedding（如果没有缓存）
  if (!word.embedding) {
    try {
      word.embedding = await getEmbedding(word.word);
      saveWords();
    } catch (err) {
      console.error(`[Game] 预计算 embedding 失败:`, err);
    }
  }

  const round: Round = {
    id: nextRoundId++,
    wordId: word.id,
    word: word.word,
    hint: word.hint,
    status: 'waiting',
    startedAt: null,
    endedAt: null,
    duration,
  };
  rounds.push(round);
  return { ...round };
}

/**
 * 开始一轮游戏
 */
export function startRound(roundId: number): Round {
  const round = rounds.find(r => r.id === roundId);
  if (!round) throw new Error('轮次不存在');
  round.status = 'active';
  round.startedAt = new Date().toISOString();
  return { ...round };
}

/**
 * 结束一轮游戏
 */
export function endRound(roundId: number): Round {
  const round = rounds.find(r => r.id === roundId);
  if (!round) throw new Error('轮次不存在');
  round.status = 'finished';
  round.endedAt = new Date().toISOString();
  return { ...round };
}

/**
 * 获取当前轮次信息
 */
export function getRound(roundId: number): Round | null {
  const round = rounds.find(r => r.id === roundId);
  return round ? { ...round } : null;
}

/**
 * 获取当前活跃的轮次
 */
export function getActiveRound(): Round | null {
  const round = [...rounds].reverse().find(r => r.status === 'active');
  return round ? { ...round } : null;
}

/**
 * 处理一次猜测
 */
export async function processGuess(
  roundId: number,
  userId: string,
  username: string,
  avatar: string,
  guessText: string
): Promise<GuessResult> {
  const maxGuesses = parseInt(process.env.MAX_GUESSES_PER_ROUND || '3');

  // 检查猜测次数
  const count = await getGuessCount(roundId, userId);
  if (count >= maxGuesses) {
    return {
      similarity: 0,
      isCorrect: false,
      rank: -1,
      message: `本轮已用完 ${maxGuesses} 次猜测机会`,
    };
  }

  // 获取当前轮次的答案
  const round = getRound(roundId);
  if (!round || round.status !== 'active') {
    return { similarity: 0, isCorrect: false, rank: -1, message: '当前没有进行中的游戏' };
  }

  // 找到答案词的 embedding
  const answerWord = words.find(w => w.id === round.wordId);
  const answerEmbedding = answerWord?.embedding;

  // 计算相似度
  const similarity = await calculateSimilarity(guessText, round.word, answerEmbedding);
  const correct = isCorrect(similarity);

  // 写入 Redis 排行榜
  await incrementGuessCount(roundId, userId);
  const entry: LeaderboardEntry = {
    rank: roundId, // addGuess 内部用 rank 字段作为 roundId
    userId,
    username,
    avatar,
    guessText,
    score: similarity,
    timestamp: Date.now(),
  };
  await addGuess(entry);

  // 记录到内存
  const record: GuessRecord = {
    roundId,
    userId,
    username,
    avatar,
    guessText,
    similarity,
    isCorrect: correct,
    createdAt: new Date().toISOString(),
  };
  guesses.push(record);
  saveGuesses();

  // 获取排名
  const leaderboard = await getLeaderboard(roundId);
  const userRank = leaderboard.findIndex(e => e.userId === userId) + 1;

  return {
    similarity: Math.round(similarity * 1000) / 1000,
    isCorrect: correct,
    rank: userRank > 0 ? userRank : leaderboard.length,
    message: correct
      ? `猜对了！相似度 ${(similarity * 100).toFixed(1)}%`
      : `相似度 ${(similarity * 100).toFixed(1)}%，继续加油！`,
  };
}

/**
 * 获取词库列表
 */
export function listWords(category?: string, difficulty?: number): Word[] {
  let result = words;
  if (category) result = result.filter(w => w.category === category);
  if (difficulty) result = result.filter(w => w.difficulty === difficulty);
  return result.map(({ embedding: _, ...rest }) => rest as Word);
}

/**
 * 添加新词到词库
 */
export function addWord(word: string, hint: string, category: string, difficulty: number): { lastInsertRowid: number } {
  const existing = words.find(w => w.word === word);
  if (existing) throw new Error(`词语 "${word}" 已存在`);
  const id = nextWordId++;
  words.push({ id, word, hint, category, difficulty });
  saveWords();
  return { lastInsertRowid: id };
}

/**
 * 删除词
 */
export function deleteWord(wordId: number): void {
  words = words.filter(w => w.id !== wordId);
  saveWords();
}

File: game-server/app/services/game_service.py
"""
游戏轮次管理服务
内存 + JSON 文件持久化
"""
import json
import logging
import random
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

from app.config import settings
from app.services.embedding_service import (
    calculate_similarity,
    is_correct,
    get_embedding,
    load_cached_embeddings,
)
from app.services.rank_service import (
    add_guess,
    get_leaderboard,
    get_guess_count,
    increment_guess_count,
    LeaderboardEntry,
)

logger = logging.getLogger(__name__)


class Word:
    def __init__(self, id: int, word: str, hint: str, category: str,
                 difficulty: int, embedding: Optional[list[float]] = None):
        self.id = id
        self.word = word
        self.hint = hint
        self.category = category
        self.difficulty = difficulty
        self.embedding = embedding

    def to_dict(self) -> dict:
        d = {
            "id": self.id, "word": self.word, "hint": self.hint,
            "category": self.category, "difficulty": self.difficulty,
        }
        if self.embedding:
            d["embedding"] = self.embedding
        return d


class Round:
    def __init__(self, id: int, word_id: int, word: str, hint: str,
                 status: str = "waiting", started_at: Optional[str] = None,
                 ended_at: Optional[str] = None, duration: int = 60):
        self.id = id
        self.word_id = word_id
        self.word = word
        self.hint = hint
        self.status = status
        self.started_at = started_at
        self.ended_at = ended_at
        self.duration = duration

    def to_dict(self) -> dict:
        return {
            "id": self.id, "wordId": self.word_id, "word": self.word,
            "hint": self.hint, "status": self.status,
            "startedAt": self.started_at, "endedAt": self.ended_at,
            "duration": self.duration,
        }


# 全局状态
_words: list[Word] = []
_rounds: list[Round] = []
_guesses: list[dict] = []
_next_word_id = 1
_next_round_id = 1


def _data_dir() -> Path:
    settings.DATA_DIR.mkdir(parents=True, exist_ok=True)
    return settings.DATA_DIR


def _load_words() -> None:
    global _words, _next_word_id
    words_file = _data_dir() / "words.json"
    if words_file.exists():
        try:
            data = json.loads(words_file.read_text(encoding="utf-8"))
            _words = [
                Word(
                    id=w["id"], word=w["word"], hint=w["hint"],
                    category=w["category"], difficulty=w["difficulty"],
                    embedding=w.get("embedding"),
                )
                for w in data.get("words", [])
            ]
            _next_word_id = data.get("nextId", max((w.id for w in _words), default=0) + 1)
        except Exception as e:
            logger.error(f"加载词库失败: {e}")
            _words = []


def _save_words() -> None:
    words_file = _data_dir() / "words.json"
    data = {
        "words": [w.to_dict() for w in _words],
        "nextId": _next_word_id,
    }
    words_file.write_text(json.dumps(data, ensure_ascii=False, indent=2), encoding="utf-8")


def _load_guesses() -> None:
    global _guesses
    guesses_file = _data_dir() / "guesses.json"
    if guesses_file.exists():
        try:
            _guesses = json.loads(guesses_file.read_text(encoding="utf-8"))
        except Exception:
            _guesses = []


def _save_guesses() -> None:
    guesses_file = _data_dir() / "guesses.json"
    guesses_file.write_text(json.dumps(_guesses, ensure_ascii=False), encoding="utf-8")


def init_game() -> None:
    """初始化游戏"""
    _load_words()
    _load_guesses()
    load_cached_embeddings([w.to_dict() for w in _words])
    logger.info(f"游戏初始化完成，词库 {_len_words()} 个词")


def _len_words() -> int:
    return len(_words)


async def create_round(
    word_id: Optional[int] = None,
    category: Optional[str] = None,
    difficulty: Optional[int] = None,
    duration: Optional[int] = None,
) -> Round:
    """创建新一轮"""
    global _next_round_id
    dur = duration or settings.ROUND_DURATION

    if word_id:
        word = next((w for w in _words if w.id == word_id), None)
        if not word:
            raise ValueError(f"词 ID {word_id} 不存在")
    else:
        candidates = _words
        if category:
            candidates = [w for w in candidates if w.category == category]
        if difficulty:
            candidates = [w for w in candidates if w.difficulty == difficulty]
        if not candidates:
            raise ValueError("词库中没有符合条件的词")
        word = random.choice(candidates)

    # 预计算 embedding
    if not word.embedding:
        try:
            word.embedding = await get_embedding(word.word)
            _save_words()
        except Exception as e:
            logger.error(f"预计算 embedding 失败: {e}")

    r = Round(
        id=_next_round_id,
        word_id=word.id,
        word=word.word,
        hint=word.hint,
        duration=dur,
    )
    _next_round_id += 1
    _rounds.append(r)
    return r


def start_round(round_id: int) -> Round:
    """开始一轮"""
    r = next((r for r in _rounds if r.id == round_id), None)
    if not r:
        raise ValueError("轮次不存在")
    r.status = "active"
    r.started_at = datetime.now(timezone.utc).isoformat()
    return r


def end_round(round_id: int) -> Round:
    """结束一轮"""
    r = next((r for r in _rounds if r.id == round_id), None)
    if not r:
        raise ValueError("轮次不存在")
    r.status = "finished"
    r.ended_at = datetime.now(timezone.utc).isoformat()
    return r


def get_round(round_id: int) -> Optional[Round]:
    return next((r for r in _rounds if r.id == round_id), None)


def get_active_round() -> Optional[Round]:
    for r in reversed(_rounds):
        if r.status == "active":
            return r
    return None


async def process_guess(
    round_id: int,
    user_id: str,
    username: str,
    avatar: str,
    guess_text: str,
) -> dict:
    """处理猜测"""
    max_guesses = settings.MAX_GUESSES_PER_ROUND

    count = get_guess_count(round_id, user_id)
    if count >= max_guesses:
        return {
            "similarity": 0, "isCorrect": False, "rank": -1,
            "message": f"本轮已用完 {max_guesses} 次猜测机会",
        }

    r = get_round(round_id)
    if not r or r.status != "active":
        return {"similarity": 0, "isCorrect": False, "rank": -1, "message": "当前没有进行中的游戏"}

    answer_word = next((w for w in _words if w.id == r.word_id), None)
    answer_embedding = answer_word.embedding if answer_word else None

    similarity = await calculate_similarity(guess_text, r.word, answer_embedding)
    correct = is_correct(similarity)

    increment_guess_count(round_id, user_id)
    ts = int(time.time() * 1000)

    entry = LeaderboardEntry(
        rank=round_id, user_id=user_id, username=username,
        avatar=avatar, guess_text=guess_text, score=similarity, timestamp=ts,
    )
    add_guess(entry)

    record = {
        "roundId": round_id, "userId": user_id, "username": username,
        "avatar": avatar, "guessText": guess_text, "similarity": similarity,
        "isCorrect": correct, "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    _guesses.append(record)
    _save_guesses()

    leaderboard = get_leaderboard(round_id)
    user_rank = next((i + 1 for i, e in enumerate(leaderboard) if e["userId"] == user_id), len(leaderboard))

    return {
        "similarity": round(similarity, 3),
        "isCorrect": correct,
        "rank": user_rank if user_rank > 0 else len(leaderboard),
        "message": (
            f"猜对了！相似度 {similarity * 100:.1f}%" if correct
            else f"相似度 {similarity * 100:.1f}%，继续加油！"
        ),
    }


def list_words(category: Optional[str] = None, difficulty: Optional[int] = None) -> list[dict]:
    result = _words
    if category:
        result = [w for w in result if w.category == category]
    if difficulty:
        result = [w for w in result if w.difficulty == difficulty]
    return [{"id": w.id, "word": w.word, "hint": w.hint, "category": w.category, "difficulty": w.difficulty} for w in result]
